system/Database/Postgre/Builder.php

```diff
@@ -312,6 +313,138 @@ public function join(string $table, $cond, string $type = '', ?bool $escape = nu
+                }
+
+                return ''; // @codeCoverageIgnore
+            }
+
+            $updateFields = $this->QBOptions['updateFields'] ??
+                $this->updateFields($keys, false, $constraints)->QBOptions['updateFields'] ??
+                [];
+
+            $alias = $this->QBOptions['alias'] ?? '_u';
+
+            $sql = 'UPDATE ' . $this->compileIgnore('update') . $table . "\n";
+
+            $sql .= "SET\n";
+
+            $this->getFieldTypes($table);
+
+            $that = $this;
+            $sql .= implode(
+                ",\n",
+                array_map(
+                    static function ($key, $value) use ($alias, $that) {
+                        $fieldName = trim($key, '"');
+
+                        return $key . ($value instanceof RawSql ?
+                                ' = ' . $value :
+                                ' = ' . $alias . '.' . $that->castValue($fieldName, $value));
+                    },
+                    array_keys($updateFields),
+                    $updateFields
+                )
+            ) . "\n";
+
+            $sql .= "FROM (\n{:_table_:}";
+
+            $sql .= ') ' . $alias . "\n";
+
+            $sql .= 'WHERE ' . implode(
+                ' AND ',
+                array_map(
+                    static function ($key, $value) use ($table, $alias, $that) {
+                        if ($value instanceof RawSql && is_string($key)) {
+                            return $table . '.' . $key . ' = ' . $value;
+                        }
+
+                        if ($value instanceof RawSql) {
+                            return $value;
+                        }
+
+                        $fieldName = trim($value, '"');
+
+                        return $table . '.' . $value . ' = ' . $alias . '.' . $that->castValue($fieldName, $value);
+                    },
+                    array_keys($constraints),
+                    $constraints
+                )
+            );
+
+            $this->QBOptions['sql'] = $sql;
+        }
+
+        if (isset($this->QBOptions['setQueryAsData'])) {
+            $data = $this->QBOptions['setQueryAsData'];
+        } else {
+            $data = implode(
+                " UNION ALL\n",
+                array_map(
+                    static fn ($value) => 'SELECT ' . implode(', ', array_map(
+                        static fn ($key, $index) => $index . ' ' . $key,
+                        $keys,
+                        $value
+                    )),
+                    $values
+                )
+            ) . "\n";
+        }
+
+        return str_replace('{:_table_:}', $data, $sql);
+    }
+
+    /**
+     * Returns cast value.
+     *
+     * @param float|int|string $value Escaped value
+     */
+    private function castValue(string $fieldName, $value): string
+    {
+        if (! isset($this->QBOptions['fieldTypes'])) {
+            throw new LogicException(
```

Comment:
I'm not sure I like throwing an exception here. I think rather than an exception if fieldTypes aren't defined then getFieldTypes() should be called. This would require the additional parameter $table though. The way it is the method is dependent on another method being called to get some data. If you don't want to add $table as another parameter then you could add $fieldTypes. 

You also would need to pass $value to the method as you could simply append ::TEXT or empty string.

```php
$value . $that->addType($fieldName, $fieldTypes);
```

Reply:
On second thought might should include value because with other DBMS as well as Postgre the format is: 

`CAST ('100' AS INTEGER)`

So: 
```php
$that->castValue($table, $fieldName, $value);
```

Reply:
```php
    private function castValue(string $table, string $fieldName, $value): string
    {
        if (! isset($this->QBOptions['fieldTypes'])) {
            $this->QBOptions['fieldTypes'] = $this->getFieldTypes($table);
        }

        $type = $this->QBOptions['fieldTypes'][$fieldName] ?? null;

        return ($type === null) ? $value : $value . '::' . strtoupper($type);
        // return ($type === null) ? $value : 'CAST(' . $value . ' AS ' . strtoupper($type) . ')';
    }
```

Reply:
```php
$that->castValue($table, $fieldName, '2024-01-01'); // CAST('2024-01-01' AS DATE)

$that->castValue($table, $fieldName, $alias . $fieldName); // CAST(_u.created_date AS DATE)
```

Reply:
You could still use the ::DATE style but in case this was ever implemented with other DBMS then it would work with this format.

Reply:
call it expression:
```php
$that->castValue($table, $fieldName, $expression);
```

Reply:
```php
$that->cast($type, $expression);

$that->cast($that->getType($table, $fieldName), $expression);

```

Reply:
Changed. https://github.com/codeigniter4/CodeIgniter4/pull/8439/commits/6d4ce25435ef1aae1c3c9bfeaf23ec5399bc77c0

Reply:
I made some updates to last post.. see what ya think

Reply:
`cast($expression, $type)` makes sense. But we can't add API in a patch version.

Changed. 92cc84e1ff1ac91c596bed6d511633cc70c5457b